### PR TITLE
Add offboard path tracking for fixedwing vehicles using NPFG

### PR DESCRIPTION
**Pull request by Jaeyoung-Lim**
**Describe problem solved by this pull request**
Fixedwing vehicles have mainly been controlled using waypoints. However, this makes it challenging to operate fixedwing vehicles close to obstacles/terrain since it is hard to predict the trajectory of the vehicle without executing the mission. 

Given that Fixedwing vehicles are more influenced by wind compared to multirotors, it is better for the vehicle to track a given **path** rather than a time stamped **trajectory**. With NPFG available in PX4, we can now precisely track arbitrary paths.

**Describe your solution**
This PR adds a path tracking interface using offboard mode for fixedwing vehicles. The vehicle can follow arbitrary paths by passing a. a position setpoint, which should be the closest point on the path b. a velocity vector, which should be the path tangent of the desired path. c. The curvature of the path. 

- Enabled NPFG by default for the believer SITL Gazebo model
- The image below shows this interface integrated into a motion primitive planner ( yellow lines are the reference path and blue is the vehicle trajectory) 

142743574-9aaa952b-c162-4655-808f-a0dbd4776376

**Test data / coverage**
- WIP testing with upstream

**Additional context**
- NPFG needs to be enabled for this usecase https://github.com/PX4/PX4-Autopilot/pull/18810

### CODE REVIEW

File: src/lib/npfg/npfg.hpp [hunk 1]
```diff
@@ -248,6 +248,22 @@ class NPFG
 			    float radius, int8_t loiter_direction, const matrix::Vector2f &ground_vel,
 			    const matrix::Vector2f &wind_vel);
 
+	/*
+	 * Loitering (unlimited) logic. Takes loiter center, radius, and direction and
```
Comment (tstastny) [inline src/lib/npfg/npfg.hpp:252]: the method documentation is still TODO

File: src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp [hunk 2]
```diff
@@ -1182,6 +1182,16 @@ FixedwingPositionControl::control_auto_position(const hrt_abstime &now, const fl
 	if (_param_fw_use_npfg.get()) {
 		_npfg.setAirspeedNom(target_airspeed * _eas2tas);
 		_npfg.setAirspeedMax(_param_fw_airspd_max.get() * _eas2tas);
+
+		if (_control_mode.flag_control_offboard_enabled && PX4_ISFINITE(pos_sp_curr.vx) && PX4_ISFINITE(pos_sp_curr.vy)) {
+			// Navigate directly on position setpoint and path tangent
+			matrix::Vector2f velocity_2d(pos_sp_curr.vx, pos_sp_curr.vy);
+			_npfg.navigatePathTangent(curr_pos, curr_wp, velocity_2d.normalized(), ground_speed, _wind_vel, pos_sp_curr.yawspeed);
```
Comment (tstastny) [inline src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp:1189]: maybe good to make a comment here noting that yawspeed is hijacked for path curvature - not sure if there should be some additional documentation in the position setpoint msg about this alternate offboard use case.. or if we leave it for now until a cleaner path interface can be made

**Review by tstastny — CHANGES_REQUESTED** [review]
@Jaeyoung-Lim thanks for getting the conversation rolling on more generalized path following interfaces. a few comments below for this one..

File: src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp [hunk 3]
```diff
@@ -1182,6 +1182,16 @@ FixedwingPositionControl::control_auto_position(const hrt_abstime &now, const fl
 	if (_param_fw_use_npfg.get()) {
 		_npfg.setAirspeedNom(target_airspeed * _eas2tas);
 		_npfg.setAirspeedMax(_param_fw_airspd_max.get() * _eas2tas);
+
+		if (_control_mode.flag_control_offboard_enabled && PX4_ISFINITE(pos_sp_curr.vx) && PX4_ISFINITE(pos_sp_curr.vy)) {
```
[the remark this answers is not in the document]
Comment (Jaeyoung-Lim) [reply src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp:1186]: Yes, so getting here means that it is already a valid position setpoint. It is checking whether it should use the path tangent in addition to a waypoint, or just treat it as a normal waypoint

File: src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp (hunk 2 above)
Comment (Jaeyoung-Lim) [reply src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp:1189]: Ah, I forgot about the hijacking I did. Should we try and use the acceleration setpoint? I can also remove it from the PR for now

File: src/lib/npfg/npfg.cpp [hunk 4]
```diff
@@ -617,6 +617,25 @@ void NPFG::navigateLoiter(const Vector2d &loiter_center, const Vector2d &vehicle
...
+			       const matrix::Vector2f &ground_vel, const matrix::Vector2f &wind_vel, const float &curvature)
+{
+	path_type_loiter_ = false;
+
+	// set unit tangent directly
+	unit_path_tangent_ = tangent_setpoint.normalized();
+
+	// closest point to vehicle
+	matrix::Vector2f error_vector = getLocalPlanarVector(position_setpoint, vehicle_pos);
+	signed_track_error_ = cross2D(unit_path_tangent_, error_vector);
```
[the remark this answers is not in the document]
Comment (Jaeyoung-Lim) [reply src/lib/npfg/npfg.cpp:632]: @tstastny Hmm good point. I think anyway we can't trust anyone that they will always be sending the closest point. would there be a better way to handle this? what is the worst that can happen if an inaccurate "closest" point is being passed?

File: src/lib/npfg/npfg.hpp (hunk 1 above)
Comment (Jaeyoung-Lim) [reply src/lib/npfg/npfg.hpp]: Fixed!

File: src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp (hunk 2 above)
Comment (Jaeyoung-Lim) [reply src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp]: This has been removed now. Curvature is passed as acceleration setpoints

File: src/lib/npfg/npfg.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (tstastny) [reply src/lib/npfg/npfg.cpp:632]: fair enough

File: src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp [hunk 5]
```diff
@@ -1191,6 +1191,18 @@ FixedwingPositionControl::control_auto_position(const hrt_abstime &now, const fl
...
+			matrix::Vector2f velocity_2d(pos_sp_curr.vx, pos_sp_curr.vy);
+			float curvature = PX4_ISFINITE(_pos_sp_triplet.current.loiter_radius) ? 1 / _pos_sp_triplet.current.loiter_radius :
+					  0.0f;
+			_npfg.navigatePathTangent(curr_pos, curr_wp, velocity_2d.normalized(), ground_speed, _wind_vel, curvature);
+
+		} else {
+			_npfg.navigateWaypoints(prev_wp, curr_wp, curr_pos, ground_speed, _wind_vel);
+		}
+
 		_npfg.navigateWaypoints(prev_wp, curr_wp, curr_pos, get_nav_speed_2d(ground_speed), _wind_vel);
```
Comment (tstastny) [inline src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp:1206]: Oops. I missed this on first pass.. this is actually overwriting your path tangent logic.. right?

**Review by tstastny — CHANGES_REQUESTED** [review]

File: src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp (hunk 5 above)
Comment (tstastny) [reply src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp:1206]: also should keep the use of `get_nav_speed_2d()` in the else statement ground_speed argument

Comment (Jaeyoung-Lim) [reply src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp]: Thanks for spotting this, I think this slipped in during a rebase!

**Comment by Jaeyoung-Lim** [comment]
Offboard path tracking working now! 

offboard_pathtracking

**Comment by Jaeyoung-Lim** [comment]
@tstastny Thanks for the review!
@sfuhrer  @RomanBapst This should be ready to go now, and it would help on moving forward https://github.com/PX4/PX4-Autopilot/pull/18980

**Review by tstastny — APPROVED** [review]

**Comment by Jaeyoung-Lim** [comment]
@tstastny Thanks!

**Pull request merged**